File: Leetcode/06_Stack & Queue/032_0150_evaluate-reverse-polish-notation.go
// https://leetcode.cn/problems/evaluate-reverse-polish-notation/description/
// m
package main

import (
	"strconv"
)

type MyStack[T any] struct {
	items []T
}

func (s *MyStack[T]) isEmpty() bool {
	return len(s.items) == 0
}

func (s *MyStack[T]) push(item T) {
	s.items = append(s.items, item)
}

func (s *MyStack[T]) pop() (T, bool) {
	var zero T
	if s.isEmpty() {
		return zero, false
	}
	cur := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return cur, true
}

func (s *MyStack[T]) peek() (T, bool) {
	var zero T
	if s.isEmpty() {
		return zero, false
	}
	return s.items[len(s.items)-1], true
}

func doMath(op1, op2 int, o string) int {
	switch o {
	case "+":
		return op1 + op2
	case "-":
		return op1 - op2
	case "*":
		return op1 * op2
	case "/":
		return op1 / op2
	}
	return 0
}

func evalRPN(tokens []string) int {
	s := MyStack[int]{}
	for _, t := range tokens {
		if t == "+" || t == "-" || t == "*" || t == "/" {
			op2, _ := s.pop() // 注意顺序，1 和 2 是反的
			op1, _ := s.pop()
			s.push(doMath(op1, op2, t))
		} else {
			num, _ := strconv.Atoi(t)
			s.push(num)
		}
	}
	res, _ := s.pop()
	return res
}

File: Leetcode/06_Stack & Queue/032_0150_evaluate-reverse-polish-notation.py
class Solution(object):
    def evalRPN(self, tokens):
        s = []
        for t in tokens:
            if t not in '+-*/':
                s.append(int(t))
            else:
                op2 = s.pop()
                op1 = s.pop()
                s.append(self.doMath(op1,op2,t))
        return s.pop()
                
    def doMath(self, op1, op2, o):
        if o == '+':
            return op1 + op2
        elif o == '-':
            return op1 - op2
        elif o == '*':
            return op1 * op2
        else:
            return int(op1 / op2)

s = Solution()
print(s.evalRPN(["10","6","9","3","+","-11","*","/","*","17","+","5","+"]))